sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -652,6 +653,299 @@ case class MapObjects private(
+ *                       collection, and which is used as input for the `valueLambdaFunction`
+ * @param valueLoopIsNull the nullability of the loop variable that is used when iterating over
+ *                        the value collection, and which is used as input for the
+ *                        `valueLambdaFunction`
+ * @param valueLoopVarDataType the data type of the loop variable that is used when iterating over
+ *                             the value collection, and which is used as input for the
+ *                             `valueLambdaFunction`
+ * @param valueLambdaFunction A function that takes the `valueLoopVar` as input, and is used as
+ *                            a lambda function to handle collection elements.
+ * @param valueInputData An expression that when evaluated returns a collection object.
+ * @param tupleLoopValue the name of the loop variable that holds the tuple to be added to the
+  *                      resulting map (used only for Scala Map)
+ * @param collClass The type of the resulting collection.
+ * @param builderValue The name of the builder variable used to construct the resulting collection.
+ */
+case class CollectObjectsToMap private(
+    keyLoopValue: String,
+    keyLoopIsNull: String,
+    keyLoopVarDataType: DataType,
+    keyLambdaFunction: Expression,
+    keyInputData: Expression,
+    valueLoopValue: String,
+    valueLoopIsNull: String,
+    valueLoopVarDataType: DataType,
+    valueLambdaFunction: Expression,
+    valueInputData: Expression,
+    tupleLoopValue: String,
+    collClass: Class[_],
+    builderValue: String) extends Expression with NonSQLExpression {
+
+  override def nullable: Boolean = keyInputData.nullable
+
+  override def children: Seq[Expression] =
+    keyLambdaFunction :: keyInputData :: valueLambdaFunction :: valueInputData :: Nil
+
+  override def eval(input: InternalRow): Any =
+    throw new UnsupportedOperationException("Only code-generated evaluation is supported")
+
+  override def dataType: DataType = ObjectType(collClass)
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val keyElementJavaType = ctx.javaType(keyLoopVarDataType)
+    ctx.addMutableState("boolean", keyLoopIsNull, "")
+    ctx.addMutableState(keyElementJavaType, keyLoopValue, "")
+    val genKeyInputData = keyInputData.genCode(ctx)
+    val genKeyFunction = keyLambdaFunction.genCode(ctx)
+    val valueElementJavaType = ctx.javaType(valueLoopVarDataType)
+    ctx.addMutableState("boolean", valueLoopIsNull, "")
+    ctx.addMutableState(valueElementJavaType, valueLoopValue, "")
+    val genValueInputData = valueInputData.genCode(ctx)
+    val genValueFunction = valueLambdaFunction.genCode(ctx)
+    val dataLength = ctx.freshName("dataLength")
+    val loopIndex = ctx.freshName("loopIndex")
+
+    // In RowEncoder, we use `Object` to represent Array or Seq, so we need to determine the type
+    // of input collection at runtime for this case.
```

Comment:
As mentioned earlier, I tried to make this class as generic and similar to `MapObjects` as possible so it can be used elsewhere without certain preconditions being met. Granted that getting sequences here in the future is unlikely. Should I remove it?